### PR TITLE
Report and set: mtu,speed, auto-negotiation and duplex

### PR DESCRIPTION
**Pull request by tyll**

### CODE REVIEW

File: libnmstate/nm/wired.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+from libnmstate import nmclient
+
+
+def get_info(active_connection):
+    """
+    Provides the current active values for an active connection.
+    """
+    info = {}
+    # FIXME: Maybe use
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Looks like we fetch the "desired" config, is it assured to be the actual? I was wandering the same for the other info we collect.

File: libnmstate/nm/wired.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    # active_connection.get_devices()[0].get_applied_connection(0) to get the
+    # real runtime connection...
+    try:
+        remote_connection = active_connection.get_connection()
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Seems we just use it as a "connection" and not as a "remote" one, but it is a nit.

File: libnmstate/nm/wired.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        wired_config = remote_connection.get_setting(nmclient.NM.SettingWired)
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Why not use get_setting_wired() ?

File: libnmstate/nm/wired.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        mtu = wired_config.get_property(nmclient.NM.SETTING_WIRED_MTU)
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Using wired_config.props.mtu is a bit nicer IMO.

File: libnmstate/nm/wired.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        info['mtu'] = mtu
+    except AttributeError:
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Will this pattern work well with future properties here?
There seem to be two levels to this: The existence of the wired config and the existence of the specific property.

**Comment by EdDev** [comment]
I feel bad to ask it (as I have not done well with this myself), but any chance you can add a test for the wired module?

File: libnmstate/nm/wired.py (hunk 1 above)
Comment (tyll) [reply libnmstate/nm/wired.py]: This might be something that we need to address in general. Currently we are using `dev.get_active_connection()` but from a discussion with @thom311 I understood that the active connection is only a pointer and might show to a configuration that is not yet, applied. But Using `dev.get_applied_connection(0)` would return the live configuration.

In general I think I would wish that there was a clear settings interface as a class that settings such as for the wired parameters could inherit. Then the parent class could provide an interface to the connection that would need to be changed or inspected to return the actual info. Maybe just provide the connection to the constructor and have two methods apply and show, where apply gets the nmstate to apply and show returns the partial state with the settings for the respective NM setting.

File: libnmstate/nm/wired.py (hunk 2 above)
Comment (tyll) [reply libnmstate/nm/wired.py]: In NM it is a RemoteConnection (as opposed to a SimpleConnection that `dev.get_applied_connection(0)` returns.

File: libnmstate/nm/wired.py (hunk 3 above)
Comment (tyll) [reply libnmstate/nm/wired.py]: I hope to reuse the code and make it more generic when adding more settings. Then only the argument to `get_setting()` is generic, which is a bit nicer that to do something like `getattr(remote_connection, 'get'setting_wired')()`.

File: libnmstate/nm/wired.py (hunk 4 above)
Comment (tyll) [reply libnmstate/nm/wired.py]: I did not know `wired_config.props`... Maybe @thom311 can explain what is better to be used. :-)

File: libnmstate/nm/wired.py (hunk 5 above)
Comment (tyll) [reply libnmstate/nm/wired.py]: We will need an extra try/except block for each setting to make sure that only individual settings are omitted.

**Comment by tyll** [comment]
I would like to add a test but I tried to figure out how to properly do the setting, first. Will need to address this in the week with August 13 because I will be at the Fedora Flock conference in the mean time.

File: libnmstate/nm/wired.py (hunk 3 above)
Comment (EdDev) [reply libnmstate/nm/wired.py]: I do not see how you will make it more generic this way, I only see here the need to depend on nmclient.NM.* (and require a separate mock for it), especially when for each of these you need to check specific attributes.
But I guess we will see it when you add another property.

File: libnmstate/nm/wired.py (hunk 4 above)
Comment (EdDev) [reply libnmstate/nm/wired.py]: Under the hood it should use the same calls.

**Comment by tyll** [comment]
So I rebased it to report the MTU from the device property and set it via a wired setting. However, strictly speaking getting the MTU information does now not belong in the wired module since it is not reading the wired configuration.

File: libnmstate/nm/applier.py [hunk 6]
```diff
@@ -194,4 +200,5 @@ def _build_connection_profile(iface_desired_state, base_con_profile=None):
         ovs_port_options = iface_desired_state.get('options')
         settings.append(ovs.create_port_setting(ovs_port_options))
 
-    return connection.create_profile(settings)
+    new_profile = connection.create_profile(settings)
```
Comment (EdDev) [inline libnmstate/nm/applier.py]: Why is this change needed?

File: libnmstate/nm/applier.py [hunk 7]
```diff
@@ -68,6 +69,7 @@ def prepare_edited_ifaces_configuration(ifaces_desired_state):
             cur_con_profile = connection.get_device_connection(nmdev)
         new_con_profile = _build_connection_profile(
             iface_desired_state, base_con_profile=cur_con_profile)
+
```
Comment (EdDev) [inline libnmstate/nm/applier.py]: nit: not related change

File: libnmstate/nm/wired.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from libnmstate import nmclient
+
+DEFAULT_MTU = 1500
+
+
+def create_setting(iface_state, base_con_profile):
+    mtu = iface_state.get("mtu")
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Single quote please.

File: libnmstate/nm/wired.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    if not mtu:
+        return None
+
+    if base_con_profile:
+        wired_setting = base_con_profile.get_setting_wired()
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: This is not how it works with the connection-settings, see: duplicate_settings()
I remember it was a recommendation from thaller.
Could you please re-check this?

File: libnmstate/nm/wired.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from libnmstate import nmclient
+
+DEFAULT_MTU = 1500
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Who is using this one (or: why do we need it?)?

File: libnmstate/nm/wired.py (hunk 8 above)
Comment (tyll) [reply libnmstate/nm/wired.py]: done

File: libnmstate/nm/wired.py (hunk 10 above)
Comment (tyll) [reply libnmstate/nm/wired.py]: removed it, it was leftover from the previous version

File: libnmstate/nm/applier.py (hunk 6 above)
Comment (tyll) [reply libnmstate/nm/applier.py]: it makes debugging easier because it allows to add `import pdb; pdb.set_trace()` after the assignment before the function returns.

File: libnmstate/nm/applier.py (hunk 7 above)
Comment (tyll) [reply libnmstate/nm/applier.py]: done

File: libnmstate/nm/wired.py (hunk 9 above)
Comment (tyll) [reply libnmstate/nm/wired.py]: @thom311 does this need to be changed somehow?

File: libnmstate/nm/applier.py (hunk 6 above)
Comment (EdDev) [reply libnmstate/nm/applier.py]: If you'll need to embed such lines in the code, then you can split it when you actually debug.
Anyway, with this reasoning we will have to double or triple the lines of code in any program, I prefer to be driven by readability and cleanness and not by a rare need of using a debugger. A temporary variable is nice only if it gives the reader more information (and sometime if it helps in avoiding breaking a line into two).

Comment (tyll) [reply libnmstate/nm/applier.py]: I prefer temporary variables in general because they allow to digest the code in smaller pieces. I can also remove this change if you prefer it.

Comment (EdDev) [reply libnmstate/nm/applier.py]: It is a nit, I will not fight over it.

File: libnmstate/nm/wired.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from libnmstate import nmclient
+
+
+def create_setting(iface_state, base_con_profile):
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: I would prefer to see this function do one thing and not two.
"create" should mean it is creating a new setting.
"update" or "set" can be a function for editing an existing setting.

So we would have something like this:
if base_con_profile:
    wired_setting = wired.duplicate_setting(base_con_profile)
else:
     wired_setting = wired.create_setting(...) # No need for args if there are no "defaults".
if wired_setting:
    wired.set_setting(wired_setting, speed=speed, mtu=mtu,...)

Comment (tyll) [reply libnmstate/nm/wired.py]: AFAIU the `else` is wrong because the `base_con_profile` might not contain a `wired_setting` (at least in theory, maybe it will always have one). IMHO the decision whether or not to create a setting would be better in the module to make it more clear for later refactoring in an interface to know what to do.

**Comment by EdDev** [comment]
I do not see the unit tests.

**Comment by tyll** [comment]
Fixed the commit, the unit tests should be there now.

**Comment by EdDev** [comment]
@tyll , the PR failed CI on flake8, could you please have a look? I also do not see the new wired module covered in the report, although the tests have been running.

**Comment by tyll** [comment]
I fixed the flake issues (I hope).

File: libnmstate/nm/wired.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    duplex = ethernet.get('duplex')
+    auto_negotiation = ethernet.get('auto-negotiation')
+
+    if not (mtu or speed or duplex or auto_negotiation):
+        return None
+
+    if auto_negotiation == YES:
+        auto_negotiation = True
+
+    if auto_negotiation == NO:
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: elif fits here

File: libnmstate/nm/wired.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+        wired_setting.set_property(nmclient.NM.SETTING_WIRED_AUTO_NEGOTIATE,
+                                   auto_negotiation)
+
+    return wired_setting
+
+
+# based on:
+# https://github.com/rlisagor/pynetlinux/blob/master/pynetlinux/ifconfig.py
+# https://elixir.bootlin.com/linux/v4.19-rc1/source/include/uapi/linux/ethtool.h
+def minimal_ethtool(interface):
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Could you create a dedicated module for this? ethtool.py under libnmstate would be nice.
Also, please review VDSM ethtool module, it includes socket protection (making sure it is closed):
https://github.com/oVirt/vdsm/blob/master/lib/vdsm/network/ethtool.py

File: libnmstate/nm/wired.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+    if auto_negotiation is not None:
+        wired_setting.set_property(nmclient.NM.SETTING_WIRED_AUTO_NEGOTIATE,
+                                   auto_negotiation)
+
+    return wired_setting
+
+
+# based on:
+# https://github.com/rlisagor/pynetlinux/blob/master/pynetlinux/ifconfig.py
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: I prefer you mention this in the commit message, comments in code are bad (they rot) but if we have them, they better explain the non trivial flow.

File: libnmstate/nm/wired.py [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+def get_info(device):
+    """
+    Provides the current active values for a device
+    """
+    info = {}
+
+    iface = device.get_iface()
+    try:
+        info['mtu'] = device.get_mtu()
+    # maybe all device actually have a get_mtu() method
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Please remove the comment.

File: libnmstate/nm/wired.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+    except AttributeError:
+        pass
+
+    ethernet = {}
+    try:
+        ethernet['speed'] = device.get_speed()
+    # DeviceWifi, DeviceGeneric, DeviceTun do not have a get_speed() method
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Same, please remove the comment. You can mention it in the doctext and/or commit message.

File: libnmstate/nm/wired.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+    except AttributeError:
+        pass
+
+    # Duplex state is not available via Network Manager:
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Same, please use the doctext or commit message.
You could also add the BZ as reference.

File: tests/lib/nm/wired_test.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+import pytest
+
+from lib.compat import mock
+
+from libnmstate import nm
+
+
+@pytest.fixture
+def NM_mock():
+    with mock.patch.object(nm.connection.nmclient, 'NM') as m:
```
Comment (EdDev) [inline tests/lib/nm/wired_test.py]: nm.wired.nmclient should be mocked.

File: libnmstate/nm/wired.py (hunk 13 above)
Comment (tyll) [reply libnmstate/nm/wired.py]: Do I Need `IFREQ_FORMAT = '16sPi'` - it works with `16sP` currently without the buffer length. I do not see any indication that there should be a buffer length. I added a finally statement to close the socket.

File: libnmstate/nm/wired.py (hunk 14 above)
Comment (tyll) [reply libnmstate/nm/wired.py]: The complex flow is deducted from the references, therefore it make sense to me to add this to the docstring.

File: libnmstate/nm/wired.py (hunk 17 above)
Comment (tyll) [reply libnmstate/nm/wired.py]: I added the BZ but I do not see how adding this to the commit message or doctext would improve this since the context to the line of code would be lost.

File: tests/lib/nm/wired_test.py (hunk 18 above)
Comment (tyll) [reply tests/lib/nm/wired_test.py]: fixed

File: libnmstate/nm/wired.py (hunk 17 above)
Comment (EdDev) [reply libnmstate/nm/wired.py]: How about moving all this to a function, there you can use doctext.

File: libnmstate/nm/wired.py [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+    iface = device.get_iface()
+    try:
+        info['mtu'] = int(device.get_mtu())
+    except AttributeError:
+        pass
+
+    ethernet = {}
+    try:
+        ethernet['speed'] = int(device.get_speed())
+    # DeviceWifi, DeviceGeneric, DeviceTun do not have a get_speed() method
```
Comment (EdDev) [inline libnmstate/nm/wired.py]: Somehow the comment here is gone.
I think it is clear without this comment.

Comment (tyll) [reply libnmstate/nm/wired.py]: As a sign of good faith I removed it. :-)
